rust/operator-binary/src/spark_k8s_controller.rs

```diff
@@ -368,62 +506,160 @@ fn pod_template_config_map(
+            .name(&cm_name)
+            .ownerreference_from_resource(spark_application, None, Some(true))
+            .context(ObjectMissingMetadataForOwnerRefSnafu)?
+            .with_recommended_labels(spark_application.build_recommended_labels("spark-submit"))
+            .build(),
+    );
+
+    product_logging::extend_config_map(
+        &RoleGroupRef {
+            cluster: ObjectRef::from_obj(spark_application),
+            role: String::new(),
+            role_group: String::new(),
+        },
+        vector_aggregator_address,
+        &config.logging,
+        SubmitJobContainer::SparkSubmit,
+        SubmitJobContainer::Vector,
+        &mut cm_builder,
+    )
+    .context(InvalidLoggingConfigSnafu { cm_name })?;
+
+    cm_builder.build().context(PodTemplateConfigMapSnafu)
 }
 
 #[allow(clippy::too_many_arguments)]
 fn spark_job(
     spark_application: &SparkApplication,
     spark_image: &str,
     serviceaccount: &ServiceAccount,
-    job_container: &Option<Container>,
     env: &[EnvVar],
     job_commands: &[String],
     s3conn: &Option<S3ConnectionSpec>,
     s3logdir: &Option<S3LogDir>,
 ) -> Result<Job> {
-    let mut cb =
-        ContainerBuilder::new("spark-submit").with_context(|_| IllegalContainerNameSnafu {
-            container_name: APP_NAME.to_string(),
-        })?;
+    let mut cb = ContainerBuilder::new(&SubmitJobContainer::SparkSubmit.to_string())
+        .context(IllegalContainerNameSnafu)?;
     let job_config = spark_application
         .job_config()
         .context(FailedToResolveConfigSnafu)?;
 
+    let log_config_map = if let Some(ContainerLogConfig {
+        choice:
+            Some(ContainerLogConfigChoice::Custom(CustomContainerLogConfig {
+                custom: ConfigMapLogConfig { config_map },
+            })),
+    }) = job_config
+        .logging
+        .containers
+        .get(&SubmitJobContainer::SparkSubmit)
+    {
+        config_map.into()
+    } else {
+        spark_application.submit_job_config_map_name()
+    };
+
+    let mut args = vec![job_commands.join(" ")];
+    if job_config.logging.enable_vector_agent {
+        // Wait for Vector to gather the logs.
+        args.push("sleep 10".into());
+        args.push(shutdown_vector_command(VOLUME_MOUNT_PATH_LOG));
+    }
+
     cb.image(spark_image)
-        .command(vec!["/bin/sh".to_string()])
+        .command(vec!["/bin/bash".to_string(), "-c".to_string()])
+        .args(vec![args.join(" && ")])
         .resources(job_config.resources.into())
-        .args(vec!["-c".to_string(), job_commands.join(" ")])
         .add_volume_mounts(spark_application.spark_job_volume_mounts(s3conn, s3logdir))
         .add_env_vars(env.to_vec())
+        .add_env_var(
+            "SPARK_SUBMIT_OPTS",
+            format!(
+                "-cp /stackable/spark/extra-jars/*:/stackable/spark/jars/* \
+                -Dlog4j.configurationFile={VOLUME_MOUNT_PATH_LOG_CONFIG}/{LOG4J2_CONFIG_FILE}"
+            ),
+        )
```

Comment:
The `env` is coming from `self.spec.env`. This could possibly overwrite `SPARK_SUBMIT_OPTS` options in there?

Reply:
Yes, it could - SPARK_SUBMIT_OPTS seems to be [undocumented](https://issues.apache.org/jira/browse/SPARK-30557) and spark.driver.extraJavaOptions and spark.executor.extraJavaOptions are preferred. However, extraJavaOptions are intended to be set by the user. We can use defaultJavaOptions instead (have not tested it) as that will be prepended to anything the user sets (see comment above).

Reply:
These are the Java options for the Spark submit job. Neither `spark.driver.defaultJavaOptions` nor `spark.executor.defaultJavaOptions` are applied in cluster mode and there is no `spark.submit.defaultJavaOptions`. As far as I can see, the only possibility to set these options is to use the `SPARK_SUBMIT_OPTS` environment variable (see https://github.com/apache/spark/blob/v3.3.0/launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java#L254-L302). It is unfortunate that this variable is undocumented and therefore not part of the public API but the logging integration test would break if this is ever changed.

Reply:
OK, looks like that is the only possibility. `self.spec.env` isn't documented by us (probably an oversight) but I'd like to leave it that way as the user shouldn't have to think about the submit job.